File: src/__tests__/verify.test.ts
import { describe, it, expect } from 'vitest';
import { secp256k1 } from '@noble/curves/secp256k1';
import { ed25519 } from '@noble/curves/ed25519';
import bs58 from 'bs58';
import { verifyEvm, eip191Digest, addressFromPublicKey } from '../evm/verify.js';
import { verifySolana } from '../solana/verify.js';
import { verifyProof } from '../verify.js';
import { buildSiwxMessage } from '../caip122.js';
import { newChallenge } from '../nonce.js';
import { bytesToHex, bytesToBase64, utf8ToBytes } from '../bytes.js';
import type { SignedProof } from '../types.js';

// --- test signers (produce real signatures the verifier must accept) ---

function evmSign(message: string) {
  const priv = secp256k1.utils.randomPrivateKey();
  const pub = secp256k1.getPublicKey(priv, false);
  const address = addressFromPublicKey(pub);
  const sig = secp256k1.sign(eip191Digest(message), priv);
  const full = new Uint8Array(65);
  full.set(sig.toCompactRawBytes(), 0);
  full[64] = (sig.recovery ?? 0) + 27;
  return { address, signature: '0x' + bytesToHex(full) };
}

function solanaSign(message: string) {
  const priv = ed25519.utils.randomPrivateKey();
  const pub = ed25519.getPublicKey(priv);
  const address = bs58.encode(pub);
  const signature = bytesToBase64(ed25519.sign(utf8ToBytes(message), priv));
  return { address, signature };
}

describe('EVM EIP-191 verify', () => {
  it('accepts a valid signature', () => {
    const message = 'hello hanzo';
    const { address, signature } = evmSign(message);
    expect(verifyEvm(message, signature, address)).toBe(true);
  });
  it('is case-insensitive on the address', () => {
    const message = 'hello';
    const { address, signature } = evmSign(message);
    expect(verifyEvm(message, signature, address.toUpperCase().replace('0X', '0x'))).toBe(true);
  });
  it('rejects a tampered message', () => {
    const { address, signature } = evmSign('original');
    expect(verifyEvm('tampered', signature, address)).toBe(false);
  });
  it('rejects a wrong address', () => {
    const { signature } = evmSign('m');
    expect(verifyEvm('m', signature, '0x0000000000000000000000000000000000000000')).toBe(false);
  });
});

describe('Solana ed25519 verify', () => {
  it('accepts a valid signature', () => {
    const message = 'hello solana';
    const { address, signature } = solanaSign(message);
    expect(verifySolana(message, signature, address)).toBe(true);
  });
  it('rejects a tampered message', () => {
    const { address, signature } = solanaSign('original');
    expect(verifySolana('tampered', signature, address)).toBe(false);
  });
  it('rejects a wrong address', () => {
    const { signature } = solanaSign('m');
    const other = bs58.encode(ed25519.getPublicKey(ed25519.utils.randomPrivateKey()));
    expect(verifySolana('m', signature, other)).toBe(false);
  });
});

describe('verifyProof end-to-end', () => {
  const now = 1_700_000_000_000;
  const base = { domain: 'hanzo.id', uri: 'https://hanzo.id/login', nonce: 'abc12345', now };

  it('accepts a fresh EVM proof', () => {
    const challenge = newChallenge(base);
    // EVM: derive the address from the key, embed it in the message, then sign.
    const priv = secp256k1.utils.randomPrivateKey();
    const address = addressFromPublicKey(secp256k1.getPublicKey(priv, false));
    const message = buildSiwxMessage({ challenge, address, chain: 'evm' });
    const sig = secp256k1.sign(eip191Digest(message), priv);
    const full = new Uint8Array(65);
    full.set(sig.toCompactRawBytes(), 0);
    full[64] = (sig.recovery ?? 0) + 27;
    const proof: SignedProof = {
      chain: 'evm', scheme: 'secp256k1-eip191', address, message,
      signature: '0x' + bytesToHex(full),
    };
    const res = verifyProof(proof, { domain: 'hanzo.id', nonce: 'abc12345', now });
    expect(res.ok).toBe(true);
    expect(res.address).toBe(address);
  });

  it('accepts a fresh Solana proof', () => {
    const challenge = newChallenge(base);
    // address is the pubkey; sign the message that embeds that address
    const priv = ed25519.utils.randomPrivateKey();
    const address = bs58.encode(ed25519.getPublicKey(priv));
    const message = buildSiwxMessage({ challenge, address, chain: 'solana' });
    const signature = bytesToBase64(ed25519.sign(utf8ToBytes(message), priv));
    const proof: SignedProof = { chain: 'solana', scheme: 'ed25519', address, message, signature };
    expect(verifyProof(proof, { domain: 'hanzo.id', nonce: 'abc12345', now }).ok).toBe(true);
  });

  it('rejects wrong nonce / domain', () => {
    const challenge = newChallenge(base);
    const address = bs58.encode(ed25519.getPublicKey(ed25519.utils.randomPrivateKey()));
    const message = buildSiwxMessage({ challenge, address, chain: 'solana' });
    const proof: SignedProof = { chain: 'solana', scheme: 'ed25519', address, message, signature: 'AAAA' };
    expect(verifyProof(proof, { domain: 'hanzo.id', nonce: 'WRONG', now }).reason).toBe('nonce-mismatch');
    expect(verifyProof(proof, { domain: 'evil.com', nonce: 'abc12345', now }).reason).toBe('domain-mismatch');
  });

  it('rejects an expired proof', () => {
    const challenge = newChallenge({ ...base, ttlSeconds: 60 });
    const priv = ed25519.utils.randomPrivateKey();
    const address = bs58.encode(ed25519.getPublicKey(priv));
    const message = buildSiwxMessage({ challenge, address, chain: 'solana' });
    const signature = bytesToBase64(ed25519.sign(utf8ToBytes(message), priv));
    const proof: SignedProof = { chain: 'solana', scheme: 'ed25519', address, message, signature };
    // now is 1h after issuance, well past 60s ttl + skew
    const res = verifyProof(proof, { domain: 'hanzo.id', nonce: 'abc12345', now: now + 3_600_000 });
    expect(res.reason).toBe('expired');
  });

  it('fails closed on an unknown scheme', () => {
    const challenge = newChallenge(base);
    const message = buildSiwxMessage({ challenge, address: 'rXYZ', chain: 'xrp' });
    const proof = { chain: 'xrp', scheme: 'totally-unknown', address: 'rXYZ', message, signature: '00' } as unknown as SignedProof;
    expect(verifyProof(proof, { domain: 'hanzo.id', nonce: 'abc12345', now }).reason).toBe('unsupported-scheme');
  });

  it('fails closed (bad-signature) on a wired-but-unverifiable proof', () => {
    const challenge = newChallenge(base);
    const message = buildSiwxMessage({ challenge, address: 'rXYZ', chain: 'xrp' });
    const proof: SignedProof = { chain: 'xrp', scheme: 'secp256k1-xrpl', address: 'rXYZ', message, signature: '00' };
    expect(verifyProof(proof, { domain: 'hanzo.id', nonce: 'abc12345', now }).ok).toBe(false);
  });
});

File: go/walletconnect/verify_test.go
package walletconnect

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"testing"
	"time"

	luxcrypto "github.com/luxfi/crypto"
	"github.com/mr-tron/base58"
)

func nowTime() time.Time { return time.UnixMilli(fixedNow) }

func baseChallenge() LoginChallenge {
	return newChallenge(challengeOpts{
		domain: "hanzo.id",
		uri:    "https://hanzo.id/login",
		nonce:  "abc12345",
		nowMs:  fixedNow,
	})
}

func TestVerifyProof_AcceptsFreshEVM(t *testing.T) {
	c := baseChallenge()
	priv, err := luxcrypto.GenerateKey()
	if err != nil {
		t.Fatalf("GenerateKey: %v", err)
	}
	address := AddressFromPublicKey(luxcrypto.FromECDSAPub(&priv.PublicKey))
	msg := mustBuild(t, BuildParams{Challenge: c, Address: address, Chain: ChainEVM})
	sig, err := luxcrypto.Sign(EIP191Digest(msg), priv)
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	sig[64] += 27
	proof := Proof{
		Chain:     ChainEVM,
		Scheme:    SchemeSecp256k1EIP191,
		Address:   address,
		Message:   msg,
		Signature: "0x" + hex.EncodeToString(sig),
	}
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if !res.OK {
		t.Fatalf("expected OK, got reason %q", res.Reason)
	}
	if res.Address != address {
		t.Errorf("address = %q want %q", res.Address, address)
	}
	if res.Chain != ChainEVM {
		t.Errorf("chain = %q", res.Chain)
	}
}

func freshSolanaProof(t *testing.T, c LoginChallenge) (Proof, string) {
	t.Helper()
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("GenerateKey: %v", err)
	}
	address := base58.Encode(pub)
	msg := mustBuild(t, BuildParams{Challenge: c, Address: address, Chain: ChainSolana})
	sig := base64.StdEncoding.EncodeToString(ed25519.Sign(priv, []byte(msg)))
	return Proof{Chain: ChainSolana, Scheme: SchemeEd25519, Address: address, Message: msg, Signature: sig}, address
}

func TestVerifyProof_AcceptsFreshSolana(t *testing.T) {
	proof, _ := freshSolanaProof(t, baseChallenge())
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if !res.OK {
		t.Fatalf("expected OK, got reason %q", res.Reason)
	}
}

func TestVerifyProof_RejectsWrongNonceAndDomain(t *testing.T) {
	// Address must match the message; signature need not verify because the
	// nonce/domain checks run before crypto. Build a real signer to satisfy the
	// address binding, then assert the pre-crypto reason.
	proof, _ := freshSolanaProof(t, baseChallenge())

	if r := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "WRONG", Now: nowTime()}).Reason; r != ReasonNonceMismatch {
		t.Errorf("nonce: reason = %q want %q", r, ReasonNonceMismatch)
	}
	if r := VerifyProof(proof, Expectation{Domain: "evil.com", Nonce: "abc12345", Now: nowTime()}).Reason; r != ReasonDomainMismatch {
		t.Errorf("domain: reason = %q want %q", r, ReasonDomainMismatch)
	}
}

func TestVerifyProof_RejectsExpired(t *testing.T) {
	c := newChallenge(challengeOpts{
		domain:     "hanzo.id",
		uri:        "https://hanzo.id/login",
		nonce:      "abc12345",
		nowMs:      fixedNow,
		ttlSeconds: 60,
	})
	proof, _ := freshSolanaProof(t, c)
	// now is 1h after issuance, well past 60s ttl + default 5m skew.
	res := VerifyProof(proof, Expectation{
		Domain: "hanzo.id",
		Nonce:  "abc12345",
		Now:    time.UnixMilli(fixedNow + 3_600_000),
	})
	if res.Reason != ReasonExpired {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonExpired)
	}
}

func TestVerifyProof_RejectsNotYetValid(t *testing.T) {
	// notBefore far in the future => not-yet-valid.
	c := baseChallenge()
	nbf := isoFromEpochMs(fixedNow + 3_600_000)
	c.NotBefore = &nbf
	proof, _ := freshSolanaProof(t, c)
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if res.Reason != ReasonNotYetValid {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonNotYetValid)
	}
}

func TestVerifyProof_RejectsAddressMismatch(t *testing.T) {
	// Message embeds a different address than the proof claims.
	proof, _ := freshSolanaProof(t, baseChallenge())
	proof.Address = randSolanaAddress(t)
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if res.Reason != ReasonAddressMismatch {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonAddressMismatch)
	}
}

func TestVerifyProof_MalformedMessage(t *testing.T) {
	proof := Proof{Chain: ChainEVM, Scheme: SchemeSecp256k1EIP191, Address: "0xabc", Message: "garbage", Signature: "0x00"}
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if res.Reason != ReasonMalformedMessage {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonMalformedMessage)
	}
}

func TestVerifyProof_FailsClosedUnknownScheme(t *testing.T) {
	c := baseChallenge()
	msg := mustBuild(t, BuildParams{Challenge: c, Address: "rXYZ", Chain: ChainXRP})
	proof := Proof{Chain: ChainXRP, Scheme: SignatureScheme("totally-unknown"), Address: "rXYZ", Message: msg, Signature: "00"}
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if res.Reason != ReasonUnsupportedScheme {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonUnsupportedScheme)
	}
}

func TestVerifyProof_FailsClosedWiredButUnverifiable(t *testing.T) {
	// A known-but-stubbed scheme (XRPL) wired with junk: passes parse/domain/
	// nonce/time, then the crypto stub returns false => bad-signature.
	c := baseChallenge()
	msg := mustBuild(t, BuildParams{Challenge: c, Address: "rXYZ", Chain: ChainXRP})
	proof := Proof{Chain: ChainXRP, Scheme: SchemeSecp256k1XRPL, Address: "rXYZ", Message: msg, Signature: "00"}
	res := VerifyProof(proof, Expectation{Domain: "hanzo.id", Nonce: "abc12345", Now: nowTime()})
	if res.OK {
		t.Fatal("expected fail-closed")
	}
	if res.Reason != ReasonBadSignature {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonBadSignature)
	}
}

func TestVerifyProof_ExpectedAddressMismatch(t *testing.T) {
	// Proof verifies, but Expectation.Address pins a different address (EVM
	// case-insensitive path).
	c := baseChallenge()
	priv, err := luxcrypto.GenerateKey()
	if err != nil {
		t.Fatalf("GenerateKey: %v", err)
	}
	address := AddressFromPublicKey(luxcrypto.FromECDSAPub(&priv.PublicKey))
	msg := mustBuild(t, BuildParams{Challenge: c, Address: address, Chain: ChainEVM})
	sig, err := luxcrypto.Sign(EIP191Digest(msg), priv)
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	sig[64] += 27
	proof := Proof{Chain: ChainEVM, Scheme: SchemeSecp256k1EIP191, Address: address, Message: msg, Signature: "0x" + hex.EncodeToString(sig)}
	res := VerifyProof(proof, Expectation{
		Domain:  "hanzo.id",
		Nonce:   "abc12345",
		Address: "0x0000000000000000000000000000000000000000",
		Now:     nowTime(),
	})
	if res.Reason != ReasonAddressMismatch {
		t.Fatalf("reason = %q want %q", res.Reason, ReasonAddressMismatch)
	}
}
